File: Sources/SomnioEditor/Canvas/CanvasController.swift
import CoreGraphics
import Foundation
import simd
import SomnioCore
import SomnioScene3D
import SwiftUI

/// Stateless canvas geometry + pick dispatcher. The drag interaction layer
/// (`DragController`) builds its sessions on these primitives; the view body stays
/// focused on layout and gesture wiring.
@MainActor public enum CanvasController {
    /// Converts a SwiftUI top-left `.local` viewport point to a legacy top-left grid
    /// coordinate: unproject through the shared camera framing onto the floor plane, then
    /// floor each axis into `Int16` (the same downward rounding the 2D pixel canvas used).
    public static func gridPoint(forViewport location: CGPoint, viewportSize: CGSize, framing: EditorFraming) -> GridPoint {
        let pixel = OrthographicCameraRig.legacyPoint(
            forViewport: SIMD2<Float>(location),
            viewportSize: SIMD2<Float>(viewportSize),
            framing: framing
        )
        return GridPoint(
            x: Int16(clamping: Int(pixel.x.rounded(.down))),
            y: Int16(clamping: Int(pixel.y.rounded(.down)))
        )
    }

    /// Navigation action a canvas scroll event resolves to.
    public enum ScrollIntent: Equatable {
        case zoom(deltaY: CGFloat)
        case pan(delta: CGSize)
    }

    /// Routes a scroll event's primitives to a navigation intent: ⌘ zooms, Shift turns a
    /// mouse wheel's vertical ticks horizontal, and line-based (non-precise) pan deltas are
    /// scaled up so one tick moves a readable distance. Zoom deltas stay raw — they feed
    /// the game's own `PlayerZoom`, whose gain is tuned against raw scroll deltas.
    public static func scrollIntent(
        deltaX: CGFloat,
        deltaY: CGFloat,
        hasPreciseDeltas: Bool,
        commandHeld: Bool,
        shiftHeld: Bool
    ) -> ScrollIntent {
        if commandHeld {
            return .zoom(deltaY: deltaY)
        }
        let lineScale: CGFloat = hasPreciseDeltas ? 1 : 10
        var delta = CGSize(width: deltaX * lineScale, height: deltaY * lineScale)
        if shiftHeld, delta.width == 0 {
            delta = CGSize(width: delta.height, height: 0)
        }
        return .pan(delta: delta)
    }

    /// Deletes every selected record in one undo step (descending-index removal per kind,
    /// see `EditorSelection.removeAll`). No-ops while a modal overlay is presented —
    /// `FantasyModalHost` swallows pointer input only, so the canvas command handlers
    /// stay wired underneath it.
    public static func deleteSelection(
        document: SectorDocument,
        workspace: SectorWorkspace,
        undoManager: UndoManager?
    ) {
        guard workspace.presentedOverlay == nil, !workspace.selection.isEmpty else { return }
        let selections = workspace.selection
        document.mutate("Delete selection", undoManager: undoManager) { body in
            EditorSelection.removeAll(selections, from: &body)
        }
        workspace.selection = []
    }

    /// Legacy-axis delta an arrow-key nudge moves the selection by: 1 px, or the grid
    /// step (floored to 1) with Shift held. Non-arrow keys resolve to `nil` so the press
    /// stays unhandled.
    public static func nudgeDelta(key: KeyEquivalent, shiftHeld: Bool, gridStep: Int16) -> (dx: Int32, dy: Int32)? {
        let step: Int32 = shiftHeld ? Int32(max(1, gridStep)) : 1
        switch key {
        case .upArrow: return (dx: 0, dy: -step)
        case .downArrow: return (dx: 0, dy: step)
        case .leftArrow: return (dx: -step, dy: 0)
        case .rightArrow: return (dx: step, dy: 0)
        default: return nil
        }
    }

    static func selectRecord(at point: GridPoint, in body: SectorBody, tool: EditorTool) -> EditorSelection? {
        // Iterate in the order returned by `candidateSelections`; within each kind walk
        // back-to-front so the most-recently-placed record wins overlaps, matching the
        // legacy editor's selection preference.
        for selection in candidateSelections(in: body, tool: tool) {
            guard let (origin, size) = selection.bounds(in: body) else { continue }
            if contains(point: point, origin: origin, size: size) {
                return selection
            }
        }
        return nil
    }

    static func candidateSelections(in body: SectorBody, tool: EditorTool) -> [EditorSelection] {
        // NPCs come first so they win when overlapping a monster spawn (original editor's
        // tool order); the Select tool hit-tests every kind, small spawn boxes before the
        // larger portal/mask/object rects so they stay reachable under overlaps.
        let npcs = body.npcs.indices.reversed().map(EditorSelection.npc)
        let monsters = body.monsterSpawns.indices.reversed().map(EditorSelection.monsterSpawn)
        let portals = body.portals.indices.reversed().map(EditorSelection.portal)
        let masks = body.collisionMasks.indices.reversed().map(EditorSelection.mask)
        let objects = body.objects.indices.reversed().map(EditorSelection.object)
        switch tool {
        case .select: return npcs + monsters + portals + masks + objects
        case .object: return objects
        case .mask: return masks
        case .portal: return portals
        case .npc: return npcs
        case .monster: return monsters
        }
    }

    private static func contains(point: GridPoint, origin: GridPoint, size: GridSize) -> Bool {
        let px = Int32(point.x)
        let py = Int32(point.y)
        let ox = Int32(origin.x)
        let oy = Int32(origin.y)
        let w = Int32(size.width)
        let h = Int32(size.height)
        return px >= ox && px < ox + w && py >= oy && py < oy + h
    }
}

File: Web/src/editor/canvasController.ts
import type * as THREE from 'three'
import { clampToInt16 } from '@/core/geometry'
import type { GridPoint } from '@/core/geometry'
import type { Sector } from '@/core/sector'
import { floorPixelAtScreen } from './picking'
import type { ScreenPoint, ViewportSize } from './picking'
import { selectionBounds } from './selection'
import type { EditorSelection } from './selection'

/**
 * Stateless canvas geometry and pick dispatch. Floor patches are a first-class tool and record.
 */

export type EditorTool = 'select' | 'object' | 'mask' | 'portal' | 'npc' | 'monster' | 'floorPatch'

export const EDITOR_TOOLS: readonly EditorTool[] = [
  'select',
  'object',
  'mask',
  'portal',
  'npc',
  'monster',
  'floorPatch',
]

/**
 * Converts a top-left viewport point to a legacy top-left grid coordinate: unproject onto the
 * floor plane, then floor each axis into Int16 (the same downward rounding the 2D pixel
 * canvas used).
 */
export function gridPoint(
  camera: THREE.OrthographicCamera,
  viewport: ViewportSize,
  screen: ScreenPoint
): GridPoint {
  const pixel = floorPixelAtScreen(camera, viewport, screen)
  return { x: clampToInt16(Math.floor(pixel.x)), y: clampToInt16(Math.floor(pixel.y)) }
}

/**
 * Legacy-axis delta an arrow-key nudge moves the selection by: 1 px, or the grid step
 * (floored to 1) with Shift held. Non-arrow keys resolve to `undefined`.
 */
export function nudgeDelta(
  key: string,
  shiftHeld: boolean,
  gridStep: number
): { dx: number; dy: number } | undefined {
  const step = shiftHeld ? Math.max(1, gridStep) : 1
  switch (key) {
    case 'ArrowUp':
      return { dx: 0, dy: -step }
    case 'ArrowDown':
      return { dx: 0, dy: step }
    case 'ArrowLeft':
      return { dx: -step, dy: 0 }
    case 'ArrowRight':
      return { dx: step, dy: 0 }
    default:
      return undefined
  }
}

/**
 * Pick candidates in preference order, back-to-front within each kind so the most-recently-
 * placed record wins overlaps. NPCs first so small spawn boxes stay reachable under the
 * larger portal/mask/object rects; floor patches last, below masks, matching their overlay
 * layer. This ordering is why a click overlapping a mask selects the mask rather than the
 * prop underneath.
 */
export function candidateSelections(sector: Sector, tool: EditorTool): EditorSelection[] {
  const reversed = (kind: EditorSelection['kind'], count: number): EditorSelection[] =>
    Array.from({ length: count }, (_, offset) => ({ kind, index: count - 1 - offset }))
  const npcs = reversed('npc', sector.npcs.length)
  const monsters = reversed('monsterSpawn', sector.monsterSpawns.length)
  const portals = reversed('portal', sector.portals.length)
  const masks = reversed('mask', sector.collisionMasks.length)
  const objects = reversed('object', sector.objects.length)
  const floorPatches = reversed('floorPatch', sector.floorPatches.length)
  switch (tool) {
    case 'select':
      return [...npcs, ...monsters, ...portals, ...masks, ...objects, ...floorPatches]
    case 'object':
      return objects
    case 'mask':
      return masks
    case 'portal':
      return portals
    case 'npc':
      return npcs
    case 'monster':
      return monsters
    case 'floorPatch':
      return floorPatches
  }
}

export function selectRecord(
  point: GridPoint,
  sector: Sector,
  tool: EditorTool
): EditorSelection | undefined {
  for (const candidate of candidateSelections(sector, tool)) {
    const bounds = selectionBounds(candidate, sector)
    if (bounds === undefined) continue
    if (
      point.x >= bounds.origin.x &&
      point.x < bounds.origin.x + bounds.size.width &&
      point.y >= bounds.origin.y &&
      point.y < bounds.origin.y + bounds.size.height
    ) {
      return candidate
    }
  }
  return undefined
}
